crypto/src/main/kotlin/web5/sdk/crypto/Signer.kt

```diff
@@ -0,0 +1,103 @@
+package web5.sdk.crypto
+
+import com.nimbusds.jose.Payload
+import com.nimbusds.jose.jwk.JWK
+
+/**
+ * A marker interface to represent options used during signing operations.
+ *
+ * Implementations of this interface might contain properties and methods that
+ * provide specific options or metadata for signature processes in different
+ * cryptographic algorithms or signing mechanisms.
+ *
+ * ### Usage Example:
+ * Implement this interface in classes where specific signing options need to
+ * be configured and passed to signing methods.
+ *
+ * ```
+ * class MySignOptions : SignOptions {
+ *     // Implementation-specific options for signing.
+ * }
+ * ```
+ *
+ * @see Signer for usage in signing operations.
+ */
+public interface SignOptions
+
+/**
+ * A marker interface to represent options used during signature verification operations.
+ *
+ * Implementations of this interface might contain properties and methods that
+ * provide specific options or metadata for signature verification processes using
+ * different cryptographic algorithms or verification mechanisms.
+ *
+ * ### Usage Example:
+ * Implement this interface in classes where specific verification options need to
+ * be configured and passed to verification methods.
+ *
+ * ```
+ * class MyVerifyOptions : VerifyOptions {
+ *     // Implementation-specific options for signature verification.
+ * }
+ * ```
+ *
+ * @see Signer for usage in signature verification operations.
+ */
+public interface VerifyOptions
+
+/**
+ * An interface defining the contract for signing and verifying signatures on payloads.
+ *
+ * `Signer` provides a generic approach to:
+ * - Creating digital signatures on payloads ([sign])
+ * - Verifying digital signatures ([verify])
+ *
+ * Implementers of this interface should ensure mechanisms provided for
+ * signing and verifying respect the cryptographic standards necessary for
+ * secure and valid operations.
+ *
+ * ### Usage Example:
+ * ```
+ * class MySigner : Signer {
+ *     override fun sign(privateKey: JWK, payload: Payload, options: SignOptions?): String {
+ *         // Implementation-specific signing logic.
+ *     }
+ *
+ *     override fun verify(publicKey: JWK, jws: String, options: VerifyOptions?) {
+ *         // Implementation-specific verification logic.
+ *     }
+ * }
+ * ```
+ *
+ * Implementers may utilize [SignOptions] and [VerifyOptions] to allow
+ * consumers to pass in additional, implementation-specific options
+ * to influence the signing and verification processes respectively.
+ *
+ * @see SignOptions for options during signing.
+ * @see VerifyOptions for options during signature verification.
+ */
+public interface Signer {
```

Comment:
I find it strange to have a `Signer` interface which also contain a `Verify` function. Is there a name that could be more appropriate? Or would it make sense to separate into two interfaces?

Reply:
understandable. i felt that concrete implementations of sign should also provide a way to verify a signature. couldn't think of a better name. open to suggestions

Reply:
My suggestion is to split into two interfaces: `verifier` and `signer`. Separating it leaves it up to the users of the library to determine which types they want, depending on their use case. I user of the library may want to only be able to verify thinks, without caring about signing.